install.c

```diff
@@ -1,29 +1,78 @@
 #include "substdio.h"
 #include "strerr.h"
+#include "env.h"
 #include "error.h"
+#include "fifo.h"
 #include "open.h"
 #include "readwrite.h"
 #include "exit.h"
+#include "alloc.h"
+#include "str.h"
+#include "stralloc.h"
 
 extern void hier();
 
 #define FATAL "install: fatal: "
 
 int fdsourcedir = -1;
 
+static void die_nomem()
+{
+  strerr_die2sys(111,FATAL,"out of memory");
+}
+
+static int ddhome(stralloc *dd, const char *home)
```

Comment:
This routine is called for every installed unit, but the result is the ~same every time (presuming all of the supporting routines always pass `auto_qmail` as `home`).  Should it be memoized?  That could be accomplished by changing where it is called (in main, which would also require a mechanism like fdsourcedir because of the change in calling convention), by using a constructor attribute for the one-time initialization, or by having a static flag for the same.  Thoughts?

Reply:
This is true, but it would require some refactoring in hier.c, which calls all these functions with the same prefix every time. This is not needed, it would be enough to pass it once in h(), but I would leave it for the moment to keep the diff small.